### ISSUE TITLE
Neural Network Wrappers

### ISSUE THREAD

**Issue by davidmascharka** [rsokl/MyGrad#26]
This can be an ongoing discussion on what to include, but convenient wrappers for the neural network functionality would be really nice. Things like optimizers to handle the bookkeeping of all your variable updates, layers that register parameters, and so on. Think PyTorch but documented and clean (so not really PyTorch at all :stuck_out_tongue:). Feel free to add to the list below, or comment on anything. This can serve as a discussion and a board of ideas.

Not all of these are probably critical (for example, who really *needs* all those optimizers?), but I think they'd be nice to have to call this fully-fledged.

We should certainly discuss how adding these may change the design of MyGrad and keep in mind design decisions while implementing these.

## Optimizers
These should take the parameters of a model and perform optimization over those parameters with respect to some loss. Learning rate schedulers may be included under here.

- [ ] (Batch) SGD [with momentum, maybe with Nesterov]
- [ ] Adam
- [ ] Adadelta
- [ ] Adagrad
- [ ] (L-)BFGS
- [ ] rmsprop

## Convenience Layers
My thought here is to have classes that handle all of the parameter registration necessary for each layer in a network. For example, a `Conv2D` layer may take `K, R, C` and create the necessary weight `Tensor`, then register its parameters with an optimizer for easy updates.

- [ ] ConvNd (N ∈{1, 2, 3} probably at the least)
- [ ] BatchNormNd
- [ ] Dense layer
- [ ] Any activations that need a dedicated  layer (adaptive layers like PReLu)
- [ ] Recurrent layers (plain RNN, LSTM, GRU)
- [ ] Dropout?

## More Losses
Should be self-explanatory

- [x] L{1,2}
- [x] Negative log-likelihood
- [x] KL divergence

## Initializers
Very handy to be able to pass a `Tensor` to a function that will initialize it according to some method.

- [x] uniform
- [x] normal
- [x] constant
- [x] identity
- [x] [dirac ](https://arxiv.org/abs/1706.00388)
- [x] [xavier/glorot](http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf) (uniform and normal)
- [x] [he/kaiming](https://arxiv.org/abs/1502.01852) (uniform and normal)
